nx/source/services/ncm.c

```diff
@@ -1,1029 +1,468 @@
-        } *resp = r.Raw;
-
-        rc = resp->result;
-
-        if (R_SUCCEEDED(rc)) {
-            if (numEntriesTotal) *numEntriesTotal = resp->total_entries;
-            if (numEntriesWritten) *numEntriesWritten = resp->entries_written;
-        }
-    }
-    
-    return rc;
+        u32 out_entries_total;
+        u32 out_entries_written;
+    } out;
+    return serviceDispatchInOut(&db->s, 7, meta_type, out, 
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { out_keys, out_keys_size } },
+    );
 }
 
-Result ncmContentMetaDatabaseHas(NcmContentMetaDatabase* db, const NcmMetaRecord* record, bool* out) {
-    IpcCommand c;
-    ipcInitialize(&c);
+Result ncmContentMetaDatabaseHas(NcmContentMetaDatabase* db, bool* out, const NcmContentMetaKey* key) {
+    return serviceDispatchInOut(&db->s, 8, *key, *out);
+}
 
-    struct {
-        u64 magic;
-        u64 cmd_id;
-        NcmMetaRecord meta_record;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 8;
-    memcpy(&raw->meta_record, record, sizeof(NcmMetaRecord));
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
+Result ncmContentMetaDatabaseHasAll(NcmContentMetaDatabase* db, bool* out, const NcmContentMetaKey* keys, size_t keys_size) {
+    return serviceDispatchOut(&db->s, 9, *out,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_In },
+        .buffers = { { keys, keys_size } },
+    );
+}
 
-        struct {
-            u64 magic;
-            u64 result;
-            bool out;
-        } *resp = r.Raw;
+Result ncmContentMetaDatabaseGetSize(NcmContentMetaDatabase* db, u64* out_size, const NcmContentMetaKey* key) {
+    return serviceDispatchInOut(&db->s, 10, *key, *out_size);
+}
 
-        rc = resp->result;
+Result ncmContentMetaDatabaseGetRequiredSystemVersion(NcmContentMetaDatabase* db, u64* out_version, const NcmContentMetaKey* key) {
+    return serviceDispatchInOut(&db->s, 11, *key, *out_version);
+}
 
-        if (R_SUCCEEDED(rc)) {
-            if (out) *out = resp->out;
-        }
-    }
-    
-    return rc;
+Result ncmContentMetaDatabaseGetPatchId(NcmContentMetaDatabase* db, u64* out_patch_id, const NcmContentMetaKey* key) {
+    return serviceDispatchInOut(&db->s, 12, *key, *out_patch_id);
 }
 
 Result ncmContentMetaDatabaseDisableForcibly(NcmContentMetaDatabase* db) {
-    IpcCommand c;
-    ipcInitialize(&c);
-
-    struct {
-        u64 magic;
-        u64 cmd_id;
-    } *raw;
-    
-    raw = ipcPrepareHeader(&c, sizeof(*raw));
-    
-    raw->magic = SFCI_MAGIC;
-    raw->cmd_id = 13;
-    
-    Result rc = serviceIpcDispatch(&db->s);
-    if (R_SUCCEEDED(rc)) {
-        IpcParsedCommand r;
-        ipcParse(&r);
-
-        struct {
-            u64 magic;
-            u64 result;
-        } *resp = r.Raw;
+    return serviceDispatch(&db->s, 13);
```

Comment:
Use _ncmCmdNoIo, same elsewhere with the relevant cmds. Also see previous comments. https://github.com/switchbrew/libnx/pull/331/files#r328164587 https://github.com/switchbrew/libnx/pull/331/files#r328161043